lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,595 @@
+                                                                   Cloned);
+  }
+
+public:
+  ExistentialSpecializerCloner(
+      SILFunction *OrigF, SILFunction *NewF, SubstitutionMap &Subs,
+      llvm::SmallVector<ArgumentDescriptor, 4> &ArgumentDescList,
+      llvm::SmallDenseMap<int, GenericTypeParamType *> &Arg2GenericTypeMap,
+      llvm::SmallDenseMap<int, ExistentialTransformArgumentDescriptor>
+          &ExistentialArgDescriptor)
+      : SuperTy(*NewF, *OrigF, Subs), OrigF(OrigF),
+        ArgumentDescList(ArgumentDescList),
+        Arg2GenericTypeMap(Arg2GenericTypeMap),
+        ExistentialArgDescriptor(ExistentialArgDescriptor) {}
+  void cloneAndPopulateFunction();
+};
+} // end anonymous namespace
+
+/// This function will create the generic version.
+void ExistentialSpecializerCloner::cloneAndPopulateFunction() {
+  // Get the Builder and the NewF
+  SILBuilder &NewFBuilder = getBuilder();
+  SILFunction &NewF = NewFBuilder.getFunction();
+
+  auto NewFTy = NewF.getLoweredFunctionType();
+  SmallVector<SILParameterInfo, 4> params;
+  params.append(NewFTy->getParameters().begin(), NewFTy->getParameters().end());
+
+  /// Builder will have a ScopeClone with a debugscope that is inherited from
+  /// the F.
+  ScopeCloner SC(NewF);
+  auto DebugScope = SC.getOrCreateClonedScope(OrigF->getDebugScope());
+  NewFBuilder.setCurrentDebugScope(DebugScope);
+  SILOpenedArchetypesTracker OpenedArchetypesTrackerNewF(&NewF);
+  NewFBuilder.setOpenedArchetypesTracker(&OpenedArchetypesTrackerNewF);
+
+  // Create the entry basic block with the function arguments.
+  SILBasicBlock *OrigEntryBB = &*OrigF->begin();
+  SILBasicBlock *ClonedEntryBB = NewF.createBasicBlock();
+  SILModule &M = OrigF->getModule();
+  auto &Ctx = M.getASTContext();
+  llvm::SmallDenseMap<int, SILInstruction *> Arg2AllocStackMap;
+  bool MissingDestroyUse = false;
+
+  BBMap.insert(std::make_pair(OrigEntryBB, ClonedEntryBB));
+  NewFBuilder.setInsertionPoint(ClonedEntryBB);
+
+  /// Determine the location for the new init_existential_ref.
+  auto InsertLoc = OrigF->begin()->begin()->getLoc();
+  for (auto &ArgDesc : ArgumentDescList) {
+    SILArgument *OrigArg = OrigEntryBB->getArgument(ArgDesc.Index);
+    auto iter = Arg2GenericTypeMap.find(ArgDesc.Index);
+    SILArgument *NewArg = nullptr;
+    /// For all Generic Arguments.
+    if (iter != Arg2GenericTypeMap.end()) {
+      auto GenericParam = iter->second;
+      SILType GenericSILType =
+          M.Types.getLoweredType(NewF.mapTypeIntoContext(GenericParam));
+      NewArg = ClonedEntryBB->createFunctionArgument(GenericSILType);
+      NewArg->setOwnershipKind(ValueOwnershipKind(
+          M, GenericSILType, ArgDesc.Arg->getArgumentConvention()));
+      /// Determine the Conformances.
+      SmallVector<ProtocolConformanceRef, 1> NewConformances;
+      auto ContextTy =
+          NewF.mapTypeIntoContext(GenericParam->getCanonicalType());
+      auto OpenedArchetype = ContextTy->getAs<ArchetypeType>();
```

Comment:
Use castTo if you know it won't fail (you're unconditionally dereferencing it below). Is it really an opened archetype at this point?